scapy/layers/hsrp.py

```diff
@@ -19,71 +19,213 @@
-# Ref. http://www.smartnetworks.jp/2006/02/hsrp_8_hsrp_version_2.html
-##
+
 # $Log: hsrp.py,v $
 # Revision 0.2  2011/05/01 15:23:34  mrenard
 # Cleanup code
 
+# scapy.contrib.description = Hot Standby Router Protocol (HSRP)
+# scapy.contrib.status = loads
+
 """
-HSRP (Hot Standby Router Protocol): proprietary redundancy protocol for Cisco routers.  # noqa: E501
+HSRP (Hot Standby Router Protocol)
+- RFC 7348 -
+
+Proprietary redundancy protocol for Cisco routers.
+
+HSRP Version 1
+    Ref. https://tools.ietf.org/html/rfc2281
+HSRP Version 2
+    Ref. http://www.smartnetworks.jp/2006/02/hsrp_8_hsrp_version_2.html
 """
 
-from scapy.fields import ByteEnumField, ByteField, IPField, SourceIPField, \
-    StrFixedLenField, XIntField, XShortField
+import struct
+from scapy.fields import ByteEnumField, ByteField, IPField, IP6Field, \
+    SourceIPField, StrFixedLenField, XIntField, XShortField, IntField, \
+    StrField, ConditionalField, MultipleTypeField
 from scapy.packet import Packet, bind_layers, bind_bottom_up
+from scapy.compat import orb
 from scapy.layers.inet import DestIPField, UDP
 from scapy.layers.inet6 import DestIP6Field
+from scapy.layers.l2 import SourceMACField
+from scapy.utils import valid_net, valid_net6
 
 
 class HSRP(Packet):
+    """
+    HSRP version 1
+    """
     name = "HSRP"
     fields_desc = [
         ByteField("version", 0),
-        ByteEnumField("opcode", 0, {0: "Hello", 1: "Coup", 2: "Resign", 3: "Advertise"}),  # noqa: E501
-        ByteEnumField("state", 16, {0: "Initial", 1: "Learn", 2: "Listen", 4: "Speak", 8: "Standby", 16: "Active"}),  # noqa: E501
+        ByteEnumField("opcode", 0,
+                      {0: "Hello",
+                       1: "Coup",
+                       2: "Resign",
+                       3: "Advertise"}),
+        ByteEnumField("state", 16,
+                      {0: "Initial",
+                       1: "Learn",
+                       2: "Listen",
+                       4: "Speak",
+                       8: "Standby",
+                       16: "Active"}),
         ByteField("hellotime", 3),
         ByteField("holdtime", 10),
         ByteField("priority", 120),
         ByteField("group", 1),
         ByteField("reserved", 0),
-        StrFixedLenField("auth", b"cisco" + b"\00" * 3, 8),
-        IPField("virtualIP", "192.168.1.1")]
+        StrFixedLenField("auth", b"cisco" + b"\x00" * 3, 8),
+        IPField("virtualIP", "192.168.1.1")
+    ]
+
+    @classmethod
+    def dispatch_hook(cls, _pkt=None, *args, **kargs):
+        """
+        Dissects or instantiates proper version of HSRP packet
+        based on the first byte.
+        """
+        # Dissect raw HSRP packet to the proper version
+        if _pkt:
+            if orb(_pkt[0]) == 0:
+                return HSRP
+            if orb(_pkt[0]) == 1:
+                return HSRPv2
+        # Create the proper HSRP version from the constructor
+        if 'version' in kargs and kargs['version'] == 2:
```

Comment:
Ok, I removed it and instantiate packet with HSRPv2 in the unit test instead. However this part: ` if orb(_pkt[0]) == 1:
                return HSRPv2` should be kept because otherwise HSRP v2 packet won't be dissected properly when IPv4 is used. I guess this is because HSRP v1 and v2 with IPv4 are using the same port ? Without this part in `dispatch_hook` scapy will dissects HSRP v2 packet (only if IPv4) as HSRP v1 packet.

Reply:
The dispatch_hook is great for dissection, but shouldn't be necessary for building.